core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -76,9 +76,12 @@ class ReplicaFetcherThread(name: String,
 
   private def epochCache(tp: TopicPartition): LeaderEpochCache =  replicaMgr.getReplica(tp).get.epochs.get
 
-  override def shutdown(): Unit = {
-    super.shutdown()
-    leaderEndpoint.close()
+  override def initiateShutdown(): Boolean = {
+    val justShutdown = super.initiateShutdown()
+    if (justShutdown) {
+      leaderEndpoint.close()
+    }
+    justShutdown
```

Comment:
Instead of overriding initiateShutdown(), perhaps it's better to do the following?

```
  override def shutdown(): Unit = {
    initiateShutdown()
    leaderEndpoint.close()
    awaitShutdown()
  }
```

Reply:
`ReplicaFetcherThread` extends from `AbstractFetcherThread` which has a [custom implementation of shutdown](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/AbstractFetcherThread.scala#L86). I did not want to duplicate this implementation in ReplicaFetcherThread.